src/monodroid/jni/monodroid-glue.c

```diff
@@ -1278,91 +1278,249 @@ gc_is_bridge_object (MonoObject *object)
+		else
+			log_error (LOG_GC, "Missing monodroidAddReference method for %s", description);
+
+		free (description);
+		free (reffed_description);
+	}
+#endif
+
+	return success;
+}
+
+// Create a target
+static AddReferenceTarget
+target_from_mono_object (MonoObject *obj)
+{
+	AddReferenceTarget result;
+	result.is_mono_object = TRUE;
+	result.obj = obj;
+	return result;
+}
+
+// Create a target
+static AddReferenceTarget
+target_from_jobject (jobject jobj)
+{
+	AddReferenceTarget result;
+	result.is_mono_object = FALSE;
+	result.jobj = jobj;
+	return result;
+}
+
+// See note in gc_prepare_for_java_collection
+static int
+scc_get_stashed_index (MonoGCBridgeSCC *scc)
+{
+	assert (scc->num_objs < 0);
+	return -scc->num_objs - 1;
+}
+
+static void
+scc_set_stashed_index (MonoGCBridgeSCC *scc, int index)
+{
+	scc->num_objs = -index - 1;
+}
+
+// Extract the root target for an SCC. If the SCC has bridged objects, this is the first object. If not, it's stored in temporary_peers.
+static AddReferenceTarget
+target_from_scc (MonoGCBridgeSCC **sccs, int idx, JNIEnv *env, jobject temporary_peers)
+{
+	MonoGCBridgeSCC *scc = sccs [idx];
+	if (scc->num_objs > 0)
+		return target_from_mono_object (scc->objs [0]);
+
+	jobject jobj = (*env)->CallObjectMethod (env, temporary_peers, jni_arraylist_get, scc_get_stashed_index (scc));
+	return target_from_jobject (jobj);
+}
+
+// Must call this on any AddReferenceTarget returned by target_from_scc once done with it
+static void
+target_release (JNIEnv *env, AddReferenceTarget target)
+{
+	if (!target.is_mono_object)
+		(*env)->DeleteLocalRef (env, target.jobj);
+}
+
+// Add a reference between objects if both are already known to be MonoObjects which are user peers
+static mono_bool
+add_reference_mono_object (JNIEnv *env, MonoObject *obj, MonoObject *reffed_obj)
+{
+	return add_reference (env, target_from_mono_object (obj), target_from_mono_object (reffed_obj));
+}
+
 static void
 gc_prepare_for_java_collection (JNIEnv *env, int num_sccs, MonoGCBridgeSCC **sccs, int num_xrefs, MonoGCBridgeXRef *xrefs)
 {
-	MonoGCBridgeSCC *scc;
-	int i, j, ref_val;
+	/* Some SCCs might have no IGCUserPeers associated with them, so we must create one */
+	jobject temporary_peers = NULL;     // This is an ArrayList
+	int temporary_peer_count = 0;       // Number of items in temporary_peers
 
-	ref_val = 1;
-	/* add java refs for items on the list which reference each other */
-	for (i = 0; i < num_sccs; i++) {
-		scc = sccs [i];
-
-		MonoJavaGCBridgeInfo    *bridge_info = NULL;
-		/* start at the second item, ref j from j-1 */
-		for (j = 1; j < scc->num_objs; j++) {
-			bridge_info = get_gc_bridge_info_for_object (scc->objs [j-1]);
-			if (bridge_info != NULL && add_reference (env, scc->objs [j-1], bridge_info, scc->objs [j])) {
-				mono.mono_field_set_value (scc->objs [j-1], bridge_info->refs_added, &ref_val);
-			}
-		}
-		/* ref the first from the last */
+	/* Basic setup before looking at xrefs */
```

Comment:
Not a very descriptive comment. The prior comment of `add java refs for items in the list which reference each other` was more descriptive. Either remove this comment (preferred?) or provide a better one.

Related: you might want to use more "helper" functions for each case:

    for (int i = 0; i < num_sccs; i++) {
        MonoGCBridgeSCC *scc = sccs [i];
        if (scc->num_objs > 1) {
            setup_whatever_refs (...);
        } else if (scc->num_objs <= 0) {
            setup_other_refs (...);
        }
    }

Another advantage to doing this is that it forces us to create names. What does it *mean* when `scc->num_objs <= 0`? That should have some kind of name, which we can then use consistently (or in only one place, but at least we have a named construct!).